### PR TITLE
add more flexibility to hyphen separated check; explicit em-dash check

### LINKED ISSUE #49: Allow github star badges
I consider GitHub star badges for resources listed like this one

```
- [rmw](https://github.com/ros2/rmw/tree/master/rmw) - Contains the ROS middleware API ![rmw](https://img.shields.io/github/stars/ros2/rmw.svg).
```

very helpful.

Right now `awesome-lint` raises an error `List item description contains invalid markdown  awesome/list-item`...

### LINKED ISSUE #146: Error caused by two dashes in item
Awesome lint uses ` - ` to separate item link and description by default. Therefore, if item name also contains a dash, it will be considered an error. As if awesome list contributor didn't respect the guidlines?

README.md
```
- [ArcGIS location services - Postman Workspace](https://www.postman.com/arcgis-developer/workspace/arcgis-location-services) - Official Postman collections to work with the Geocoding & Search API, Routing & Directions API, Demographics & GeoEnrichment API, Data hosting and more.
```

awesome-lint
```shell
  README.md:43:126
  ✖  43:126  List item link and description must be separated with a dash  remark-lint:awesome-list-item
```

From: https://github.com/Esri/awesome-arcgis-developer/runs/5556738428?check_suite_focus=true

### PR DESCRIPTION
**Pull request by WillForan**
test to confirm #146 and #49 are resolved (though I think at least 146 was already making it through without getting flagged)

The big kludge in this pull request is allowing description text matching " - " to pass through instead of getting hit with "link and description must be separated with a dash." 

I also un-commented the TODO list-items fixture that had want-to-be valid image-before-link entries. 
They had em-dash separators instead of standard hyphens. I updated lint message to complain about that in addition to en-dashes.

Fixes #146
Fixes #49

### CODE REVIEW

File: rules/list-item.js [hunk 1]
```diff
@@ -168,11 +187,17 @@ function validateListItemDescription(description, file) {
 			return false;
 		}
 
-		if (/^\s*—/.test(prefixText)) {
-			file.message('List item link and description separated by invalid en-dash', prefix);
+		if (/^\s*[—–]/.test(prefixText)) {
```
Comment (sindresorhus) [inline rules/list-item.js:190]: I think we should use Unicode escapes for the dash characters to make it more readable.

File: rules/list-item.js [hunk 2]
```diff
@@ -168,11 +187,17 @@ function validateListItemDescription(description, file) {
...
+			file.message('List item link and description separated by invalid en-dash or em-dash', prefix);
 			return false;
 		}
 
+		// Might have image and link on left side before desciption.
+		// Assume a hyphen with spaces in the description is good enough
+		if (/ - +[A-Z]/.test(descriptionText)) {
```
Comment (sindresorhus) [inline rules/list-item.js:197]:
```suggestion
		if (/ - [A-Z]/.test(descriptionText)) {
```

File: rules/list-item.js [hunk 3]
```diff
@@ -102,12 +102,24 @@ function validateList(list, file) {
...
-		const [link, ...description] = paragraph.children;
-
-		if (link.type === 'text') {
+		if (paragraph.children[0].type === 'text') {
 			continue;
 		}
 
+		let [link, ...description] = paragraph.children;
+
+		// Might have children like {image} {text} {link} - {descrition}
```
Comment (sindresorhus) [inline rules/list-item.js:111]: Typo

File: rules/list-item.js [hunk 4]
```diff
@@ -102,12 +102,24 @@ function validateList(list, file) {
...
+		// Keep discarding elements until we find a link
+		//                                       or linkReference
+		if (!/link/.test(link.type)) {
```
Comment (sindresorhus) [inline rules/list-item.js:114]: Compare explicitly. Doesn't make sense to use regex here.

File: rules/list-item.js [hunk 5]
```diff
@@ -102,12 +102,24 @@ function validateList(list, file) {
...
+			for (let i = 0; i < description.length - 1; i++) {
```
Comment (sindresorhus) [inline rules/list-item.js:115]: Use a `for-of` loop.

File: rules/list-item.js [hunk 6]
```diff
@@ -102,12 +102,24 @@ function validateList(list, file) {
...
-		if (link.type === 'text') {
+		if (paragraph.children[0].type === 'text') {
 			continue;
 		}
 
+		let [link, ...description] = paragraph.children;
+
+		// Might have children like {image} {text} {link} - {descrition}
+		// Keep discarding elements until we find a link
+		//                                       or linkReference
```
Comment (sindresorhus) [inline rules/list-item.js:113]:
```suggestion
		// Keep discarding elements until we find a `link` or `linkReference`.
```

File: rules/list-item.js [hunk 7]
```diff
@@ -117,6 +129,13 @@ function validateList(list, file) {
 }
 
 function validateListItemLink(link, file) {
+	// NB. need remark-lint-no-undefined-references separately
+	//     to catch if this is a valid reference. here we only care that it exists
```
Comment (sindresorhus) [inline rules/list-item.js:133]:
```suggestion
	// NB. We need remark-lint-no-undefined-references separately
	// to catch if this is a valid reference. Here we only care that it exists.
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: rules/list-item.js (hunk 4 above)
Comment (WillForan) [reply rules/list-item.js:114]: besides linkReference and link, I'm not sure if there are other link items in markdown -- figured any type matching "link" should get to pass
`link.type !== "linkReference" && linkk.type !== "link"` is a mouth full :)

Comment (sindresorhus) [reply rules/list-item.js:114]: You could use `Array#includes`.

---

> besides linkReference and link, I'm not sure if there are other link items in markdown

Look at the docs for the Markdown parser.

File: rules/list-item.js (hunk 5 above)
Comment (WillForan) [reply rules/list-item.js:115]: worse, I don't think I need tow be testing link's type twice. while loop is probably better?

File: rules/list-item.js (hunk 1 above)
Comment (WillForan) [reply rules/list-item.js]: (done! but apparently didn't do this in a way github's ui understandas. sorry!)

File: rules/list-item.js [hunk 8]
```diff
@@ -117,6 +124,13 @@ function validateList(list, file) {
 }
 
 function validateListItemLink(link, file) {
+	// NB. We need remark-lint-no-undefined-references separately
+	// to catch if this is a valid reference. Here we only care that it exists.
+	if (link.type === 'linkReference') {
+		// TODO: need to test link children against listItemLinkNodeAllowList?
```
Comment (sindresorhus) [inline rules/list-item.js:130]: This needs to be resolved.

**Pull request merged**